viewer/packages/tools/src/Timeline/TimelineTool.ts

```diff
@@ -0,0 +1,203 @@
+      default:
+        assertNever(event, `Unsupported event: '${event}'`);
+    }
+  }
+
+  /**
+   * Create Key frame for the Timeline
+   * @param date - date value by Date.now() since January 1, 1970
+   */
+  public createKeyframe(date: Date): Keyframe {
+    const keyframe = new Keyframe(this._model, date);
+    this._keyframes.push(keyframe);
+    this.sortKeyframesByDates();
+
+    return keyframe;
+  }
+
+  /**
+   * Returns the keyframe at the date given, or undefined if not found.
+   * @param date
+   * @returns
+   */
+  public getKeyframeByDate(date: Date): Keyframe | undefined {
+    return this._keyframes.find(candidate => candidate.getKeyframeDate() === date);
+  }
+
+  /**
+   * Removes the Keyframe from the timeline. Does nothing if the keyframe isn't part of the timeline.
+   * @param keyframe - Keyframe to be removed from the timeline
+   */
+  public removeKeyframe(keyframe: Keyframe) {
+    if (this._keyframes.length > 0) {
+      const index = this._keyframes.findIndex(obj => obj === keyframe);
+
+      if (index > -1) {
+        this._keyframes = this._keyframes.splice(index, 1);
+      }
+    }
+  }
+
+  /**
+   * Removes the Keyframe from the Timeline
+   * @param date - Date of the Keyframe to be removed from the Timeline
+   */
+  public removeKeyframeByDate(date: Date) {
+    if (this._keyframes.length > 0) {
+      const index = this._keyframes.findIndex(obj => obj.getKeyframeDate() === date);
+
+      if (index > -1) {
+        this._keyframes = this._keyframes.splice(index, 1);
+      }
+    }
+  }
+
+  /**
+   * Starts playback of Timeline
+   * @param startDate - Keyframe date to start the Playback of Keyframes
+   * @param endDate - Keyframe date to stop the Playback of Keyframes
+   * @param totalDurationInMilliSeconds - Number of milliseconds for all Keyframe within startDate & endDate to be rendered
+   */
+  public play(startDate: Date, endDate: Date, totalDurationInMilliSeconds: number) {
+    this.stop();
+
+    const playState = { dateInMs: startDate.getTime() };
+    const to = { dateInMs: endDate.getTime() };
+    const tween = new TWEEN.Tween(playState).to(to, totalDurationInMilliSeconds);
+    let currentKeyframeIndex = -1;
+    tween.onUpdate(() => {
+      const date = new Date(playState.dateInMs);
+
+      // Forward active keyframe to last keyframe that is before current date
+      const prevIndex = currentKeyframeIndex;
+      while (
+        currentKeyframeIndex < this._keyframes.length - 1 &&
+        this._keyframes[currentKeyframeIndex + 1].getKeyframeDate().getTime() <= date.getTime()
+      ) {
+        currentKeyframeIndex++;
+      }
+
+      if (currentKeyframeIndex !== prevIndex) {
+        if (prevIndex !== -1) {
+          this._keyframes[prevIndex].deactivate();
+        }
+        this._keyframes[currentKeyframeIndex].activate();
+      }
+
+      this._events.dateChanged.fire({
+        date: new Date(Date.now()),
+        activeKeyframe: this._keyframes[currentKeyframeIndex],
```

Comment:
`currentKeyframeIndex` can be -1, needs to be handled (in this case `activeKeyframe` should be null or undefined, you decide). Unittest that this works.

Reply:
if `currentKeyframeIndex` is -1 than `activeKeyframe` will be `undefined`, but will add unit test for it

Reply:
Make sure its also reflected in docs that it can be undefined.